test/core/bad_client/bad_client.cc

```diff
@@ -62,136 +69,172 @@ static void server_setup_transport(void* ts, grpc_transport* transport) {
                               grpc_server_get_channel_args(a->server));
 }
 
-static void read_done(void* arg, grpc_error* error) {
+/* Sets the read_done event */
+static void set_read_done(void* arg, grpc_error* error) {
   gpr_event* read_done = (gpr_event*)arg;
   gpr_event_set(read_done, (void*)1);
 }
 
-void grpc_run_bad_client_test(
-    grpc_bad_client_server_side_validator server_validator,
-    grpc_bad_client_client_stream_validator client_validator,
-    const char* client_payload, size_t client_payload_length, uint32_t flags) {
-  grpc_endpoint_pair sfd;
-  thd_args a;
-  gpr_thd_id id;
-  char* hex;
-  grpc_transport* transport;
-  grpc_slice slice =
-      grpc_slice_from_copied_buffer(client_payload, client_payload_length);
-  grpc_slice_buffer outgoing;
-  grpc_closure done_write_closure;
-  grpc_core::ExecCtx exec_ctx;
-  grpc_completion_queue* shutdown_cq;
+/* shutdown client */
+static void shutdown_client(grpc_endpoint** client_fd) {
+  if (*client_fd != nullptr) {
+    grpc_endpoint_shutdown(
+        *client_fd, GRPC_ERROR_CREATE_FROM_STATIC_STRING("Forced Disconnect"));
+    grpc_endpoint_destroy(*client_fd);
+    grpc_core::ExecCtx::Get()->Flush();
+    *client_fd = nullptr;
+  }
+}
 
-  if (client_payload_length < 4 * 1024) {
-    hex = gpr_dump(client_payload, client_payload_length,
+/* Runs client side validator */
+void grpc_run_client_side_validator(grpc_bad_client_arg* arg, uint32_t flags,
```

Comment:
I realize that this was an existing problem, but it looks like flags is only used to signal shutdown. If that is the case it should be renamed to shutdown. If there are other future uses then it should be used like if (flags & FLAGS_SHUTDOWN) {...}

Reply:
Currently, there are two uses for flags.
One, to indicate that the client should disconnect before the validator runs.
Second, to indicate that the client is going to write too much data which might not finish writing even after the validator returns. This ends up requiring a forceful shutdown after validator runs, so  we can avoid a SEGFAULT.

I've updated the comments and the if condition to reflect the change.